The `JavaScript Xlang Test` failure is unrelated to this PR (the diff only touches `java/fory-core` and one docs page; all Java jobs are green).

The job fails during **Install JavaScript dependencies**: `@types/node@26.1.0` (published 2026-07-01) gets hoisted into the root `node_modules`, and its new `ffi.d.ts` uses syntax that TypeScript `^4.8.4` (the compiler version used in `javascript/`) cannot parse:

```
##[error]../../node_modules/@types/node/ffi.d.ts(94,21): error TS1139: Type parameter declaration expected.
##[error]../../node_modules/@types/node/ffi.d.ts(94,29): error TS1005: ',' expected.
```

Any run of this job started after 2026-07-01 should hit the same failure until `@types/node` is pinned (the direct dep is `^18.7.21`, but a transitive `@types/node: *` wins the hoist) or TypeScript is upgraded in the `javascript` workspace.

--- a/docs/security/deserialization.md
+++ b/docs/security/deserialization.md
@@ -164,13 +164,18 @@ For stream-backed input:
 - A stream-backed buffer may hold the full requested encoded body after that
   body has been read from the stream. It must not reserve the attacker-declared
   length before input bytes prove that length exists.
-- Stream-backed fill buffers should grow from the current proven buffer size,
-  such as by doubling current capacity, and cap only to the immediate target
-  when the next bounded growth step reaches it. A byte owner may use an
-  owner-local availability signal as a one-shot growth hint when the stream
-  implementation itself is caller-owned trusted code; if that hint is absent or
-  insufficient, the reader must fall back to bounded growth from already
-  buffered bytes. Serializers should not add their own availability branches.
+- Stream-backed fill buffers should grow geometrically from the current proven
+  buffer size, such as by doubling current capacity. Growth must not be capped
+  to the immediate fill target: for small fills the target is barely above the
+  current capacity, so cap-to-target degenerates into constant-size growth
+  steps that copy the whole buffer on every small read and make stream
+  deserialization O(n^2) overall. A byte owner may use an owner-local
+  availability signal as a one-shot growth hint when the stream implementation
+  itself is caller-owned trusted code, and may then reserve the full immediate
+  target at once while keeping at least the geometric growth step; if that hint
+  is absent or insufficient, the reader must fall back to bounded geometric
+  growth from already buffered bytes. Serializers should not add their own
+  availability branches.
 - A truncated stream should fail before allocating the final deserialized value
   and should allocate only for bytes actually read plus bounded spare capacity.
 

--- a/java/fory-core/src/main/java/org/apache/fory/io/ForyInputStream.java
+++ b/java/fory-core/src/main/java/org/apache/fory/io/ForyInputStream.java
@@ -64,7 +64,7 @@ public class ForyInputStream extends InputStream implements ForyStreamReader {
         int offset = buffer.size();
         int remainingNeeded = minFillSize - totalRead;
         long targetSize = (long) offset + remainingNeeded;
-        if (targetSize > Integer.MAX_VALUE - 8L) {
+        if (targetSize > MAX_BUFFER_SIZE) {
           throw new IndexOutOfBoundsException("Stream buffer size exceeds supported range");
         }
         if (targetSize > heapMemory.length) {
@@ -72,14 +72,17 @@ public class ForyInputStream extends InputStream implements ForyStreamReader {
           if (!checkedAvailable) {
             checkedAvailable = true;
             // Use available() only as a one-shot growth hint. It may be expensive or
-            // conservative, so failed hints fall back to bounded doubling. Final value
-            // allocation still waits for fillBuffer to complete successfully.
+            // conservative, so failed hints fall back to doubling. Grow by at least a
+            // doubling step so that repeated small fills stay amortized O(1); growing to
+            // the exact target size would copy the whole buffer on every small read,
+            // making stream deserialization O(n^2) overall.
             if (stream.available() >= remainingNeeded) {
-              newSize = (int) targetSize;
+              newSize =
+                  (int) Math.max(targetSize, ForyStreamReader.nextBufferSize(heapMemory.length));
             }
           }
           if (newSize == 0 && offset == heapMemory.length) {
-            newSize = nextBufferSize(heapMemory.length, (int) targetSize);
+            newSize = ForyStreamReader.nextBufferSize(heapMemory.length);
           }
           if (newSize != 0) {
             heapMemory = growBuffer(buffer, newSize);
@@ -111,15 +114,6 @@ public class ForyInputStream extends InputStream implements ForyStreamReader {
     buffer.initHeapBuffer(newBuffer, 0, size);
     heapMemory = newBuffer;
     return heapMemory;
-  }
-
-  private static int nextBufferSize(int size, int targetSize) {
-    long grown = size == 0 ? 1L : (long) size << 1;
-    int maxSize = Integer.MAX_VALUE - 8;
-    if (grown > maxSize) {
-      grown = maxSize;
-    }
-    return (int) Math.min(grown, targetSize);
   }
 
   @Override

--- a/java/fory-core/src/main/java/org/apache/fory/io/ForyReadableChannel.java
+++ b/java/fory-core/src/main/java/org/apache/fory/io/ForyReadableChannel.java
@@ -98,7 +98,7 @@ public class ForyReadableChannel implements ForyStreamReader, ReadableByteChanne
         int position = byteBuf.position();
         int remainingNeeded = minFillSize - totalRead;
         long targetSize = (long) position + remainingNeeded;
-        if (targetSize > Integer.MAX_VALUE) {
+        if (targetSize > MAX_BUFFER_SIZE) {
           throw new DeserializationException("Stream buffer size exceeds supported range");
         }
         if (targetSize > byteBuf.capacity()) {
@@ -107,12 +107,16 @@ public class ForyReadableChannel implements ForyStreamReader, ReadableByteChanne
             checkedSeekableRemaining = true;
             // Query exact channel remaining bytes only as a one-shot fast path. Otherwise grow
             // from bytes already buffered so truncated channels fail before reserving the body.
+            // Grow by at least a doubling step so that repeated small fills stay amortized
+            // O(1); growing to the exact target size would copy the whole buffer on every
+            // small read, making stream deserialization O(n^2) overall.
             if (seekableChannel.size() - seekableChannel.position() >= remainingNeeded) {
-              newCapacity = (int) targetSize;
+              newCapacity =
+                  (int) Math.max(targetSize, ForyStreamReader.nextBufferSize(byteBuf.capacity()));
             }
           }
           if (newCapacity == 0 && position == byteBuf.capacity()) {
-            newCapacity = nextBufferSize(byteBuf.capacity(), (int) targetSize);
+            newCapacity = ForyStreamReader.nextBufferSize(byteBuf.capacity());
           }
           if (newCapacity != 0) {
             byteBuf = growBuffer(byteBuf, memoryBuf, position, newCapacity);
@@ -149,14 +153,6 @@ public class ForyReadableChannel implements ForyStreamReader, ReadableByteChanne
     byteBuffer = newByteBuf;
     memoryBuf.initByteBuffer(newByteBuf, position);
     return newByteBuf;
-  }
-
-  private static int nextBufferSize(int oldCapacity, int targetSize) {
-    long grown = oldCapacity == 0 ? 1L : (long) oldCapacity << 1;
-    if (grown > Integer.MAX_VALUE) {
-      grown = Integer.MAX_VALUE;
-    }
-    return (int) Math.min(grown, targetSize);
   }
 
   @Override

--- a/java/fory-core/src/main/java/org/apache/fory/io/ForyStreamReader.java
+++ b/java/fory-core/src/main/java/org/apache/fory/io/ForyStreamReader.java
@@ -34,6 +34,23 @@ import org.apache.fory.memory.MemoryBuffer;
 public interface ForyStreamReader {
 
   /**
+   * Maximum buffer size to use when growing stream-backed buffers: the largest array size commonly
+   * supported by JVM implementations.
+   */
+  int MAX_BUFFER_SIZE = Integer.MAX_VALUE - 8;
+
+  /**
+   * Returns the next geometric growth step for a stream-backed buffer of the given capacity,
+   * bounded by {@link #MAX_BUFFER_SIZE}. Growth must not be capped to the immediate fill target:
+   * for small fills that degenerates into constant-size growth steps which copy the whole buffer on
+   * every small read and make stream deserialization O(n^2) overall.
+   */
+  static int nextBufferSize(int currentCapacity) {
+    long grown = currentCapacity == 0 ? 1L : (long) currentCapacity << 1;
+    return (int) Math.min(grown, MAX_BUFFER_SIZE);
+  }
+
+  /**
    * Read stream and fill the data to underlying {@link MemoryBuffer}, which is also the buffer
    * returned by {@link #getBuffer}.
    */

--- a/java/fory-core/src/test/java/org/apache/fory/StreamTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/StreamTest.java
@@ -372,6 +372,51 @@ public class StreamTest extends ForyTestBase {
   }
 
   @Test
+  public void testStreamBufferGrowthIsGeometric() throws IOException {
+    // Reading many small values from a stream must not reallocate the internal buffer
+    // on every read: exact-fit growth copies the whole buffer per small fill and makes
+    // stream deserialization O(n^2), which looks like a hang for multi-MB payloads.
+    byte[] data = new byte[1 << 16];
+    ForyInputStream input = new ForyInputStream(new ByteArrayInputStream(data), 64);
+    assertGeometricGrowth(input.getBuffer(), data.length, "stream");
+
+    Path tempFile = Files.createTempFile("geometric_growth", "data");
+    Files.write(tempFile, data);
+    try {
+      try (ForyReadableChannel heapChannel =
+          new ForyReadableChannel(Files.newByteChannel(tempFile), ByteBuffer.allocate(64))) {
+        assertGeometricGrowth(heapChannel.getBuffer(), data.length, "heap channel");
+      }
+      try (ForyReadableChannel directChannel =
+          new ForyReadableChannel(Files.newByteChannel(tempFile), ByteBuffer.allocateDirect(64))) {
+        assertGeometricGrowth(directChannel.getBuffer(), data.length, "direct channel");
+      }
+    } finally {
+      Files.delete(tempFile);
+    }
+  }
+
+  private static void assertGeometricGrowth(MemoryBuffer buffer, int numBytes, String label) {
+    int growCount = 0;
+    Object lastBacking = backingBuffer(buffer);
+    for (int i = 0; i < numBytes; i++) {
+      buffer.readByte();
+      if (backingBuffer(buffer) != lastBacking) {
+        lastBacking = backingBuffer(buffer);
+        growCount++;
+        assertTrue(
+            growCount <= 20,
+            label + " buffer must grow geometrically, but already grew " + growCount + " times");
+      }
+    }
+  }
+
+  private static Object backingBuffer(MemoryBuffer buffer) {
+    byte[] heapMemory = buffer.getHeapMemory();
+    return heapMemory != null ? heapMemory : buffer.getOffHeapBuffer();
+  }
+
+  @Test
   public void testScopedMetaShare() throws IOException {
     Fory fory =
         Fory.builder()

--- a/javascript/package.json
+++ b/javascript/package.json
@@ -15,6 +15,7 @@
   "devDependencies": {
     "@stylistic/eslint-plugin": "^1.5.1",
     "@types/js-beautify": "^1.14.3",
+    "@types/node": "18.19.130",
     "eslint": "^8.55.0",
     "jest": "^29.5.0",
     "jest-junit": "^17.0.0",

